# Module 4 Lab 2: Textbook Problem 3.1

### The Bookstore Inventory

**Name:** Mac Mckay

**Course:** CS 82A

**Repo:** https://github.com/raulmckay-blip/cs82a-portfolio

## Overview

This lab uses `allbacks.csv`, which has 15 books with their `volume` (cm³), cover `area` (cm², 0 for paperbacks), `weight` (g), and `cover` type (`hb` = hardcover, `pb` = paperback). The goal is to find the median cover type, the mean weight of all the books, and the mean weight for each cover type, then compare the two groups.

## Load the data

In [8]:
import pandas as pd

books = pd.read_csv('allbacks.csv', index_col=0)
print(books.shape)   # expect (15, 4)
books

(15, 4)


,volume,area,weight,cover
rownames,,,,
1,885,382,800,hb
2,1016,468,950,hb
3,1125,387,1050,hb
4,239,371,350,hb
5,701,371,750,hb
6,641,367,600,hb
7,1228,396,1075,hb
8,412,0,250,pb
9,953,0,700,pb


In [9]:
books['cover'].value_counts()

cover
pb    8
hb    7
Name: count, dtype: int64

## Median cover type

Cover type isn't a number, but it still has an order. Paperback is the cheaper, lower binding and hardcover is the higher one. So I set the order as `pb` < `hb`, sort all the books, and take the middle one, same as a median with numbers.

In [10]:
cover_order = pd.CategoricalDtype(['pb', 'hb'], ordered=True)
covers = books['cover'].astype(cover_order).sort_values().reset_index(drop=True)

n = len(covers)
middle = n // 2          # index 7 = the 8th book
median_cover = covers[middle]

print(list(covers))
print(f'n = {n}, middle position = {middle + 1}, median cover = {median_cover}')

['pb', 'pb', 'pb', 'pb', 'pb', 'pb', 'pb', 'pb', 'hb', 'hb', 'hb', 'hb', 'hb', 'hb', 'hb']
n = 15, middle position = 8, median cover = pb


Note: The median cover type is paperback. 

## Mean weight of all books

In [11]:
mean_weight = books['weight'].mean()
print(f'Mean weight of all books: {mean_weight:.2f} g')

Mean weight of all books: 706.67 g


Note: The average book weighs about 707 g, which is a little over 1.5 lb.

## Mean weight by cover type

In [12]:
group_means = books.groupby('cover')['weight'].mean()
print(group_means.round(2))

diff = group_means['hb'] - group_means['pb']
print(f'\nHardcovers are {diff:.2f} g heavier on average')

cover
hb    796.43
pb    628.12
Name: weight, dtype: float64

Hardcovers are 168.30 g heavier on average


Note; Hardcovers average about 796 g and paperbacks about 628 g, so hardcovers are around 168 g heavier.

## Checking book size

Before saying the cover is what makes the difference, I wanted to check if the hardcovers are just bigger books. If they were, that could explain the extra weight instead of the cover.

In [13]:
books.groupby('cover')['volume'].mean().round(1)

cover
hb    833.6
pb    855.5
Name: volume, dtype: float64

## What the difference suggests

The two groups are pretty much the same size:

Paperbacks are actually a little bigger on average (856 vs 834 cm³). Even so, hardcovers still weigh about 168 g more, so it looks like the hard cover itself adds real weight.


## By-hand check

**Paperback mean weight:**

250 + 700 + 650 + 975 + 350 + 950 + 425 + 725 = 5025

5025 / 8 = **628.125 g**, which matches what pandas got.

**Median cover type:**

If I line up all 15 books with paperbacks first, it looks like this:

pb, pb, pb, pb, pb, pb, pb, pb, hb, hb, hb, hb, hb, hb, hb

With 15 books the middle one is the 8th, since there are 7 books on each side of it. Counting over, book 8 is the last paperback, so the median cover type is **pb**.

In [14]:
pb_weights = [250, 700, 650, 975, 350, 950, 425, 725]
print('By-hand pb mean  :', sum(pb_weights) / len(pb_weights))
print('pandas pb mean   :', group_means['pb'])

By-hand pb mean  : 628.125
pandas pb mean   : 628.125


## AI Use

I used AI to help with syntax and refine wording on this lab.